frontend/model/captureLogs.js

```diff
@@ -0,0 +1,176 @@
+import sbp from '~/shared/sbp.js'
+import { CAPTURED_LOGS, SET_APP_LOGS_FILTER } from '~/frontend/utils/events.js'
+
+/*
+  giConsole/count - total logs stored
+  giConsole/lastEntry - latest log hash.
+  giConsole/markers - an array of markers at every 1000th entry.
+*/
+const ENTRIES_LIMIT = 4000
+const ENTRIES_MARKER = 1000
+
+const getMarkers = () => JSON.parse(localStorage.getItem('giConsole/markers')) || []
+
+// these are initialized at captureLogsStart()
+let isCapturing = false
+let isConsoleOveridden = false
+let lastEntry = null
+let entriesCount = null
+let appLogsFilter = []
+
+function captureLog (type, ...msg) {
+  const logEntry = `${Date.now()}_${Math.floor(Math.random() * 100000)}` // uuid
+
+  // detect when is an Error and capture it properly
+  // ex: uncaught Vue errors or custom try/catch errors.
+  msg = msg.map((m) => m instanceof Error ? (JSON.stringify(m.stack) || m) : m)
+
+  localStorage.setItem(`giConsole/${logEntry}`,
+    JSON.stringify({
+      type,
+      msg,
+      prev: lastEntry,
+      timestamp: new Date().toISOString()
+    })
+  )
+
+  lastEntry = logEntry
+  entriesCount++
+
+  localStorage.setItem('giConsole/lastEntry', logEntry)
+  localStorage.setItem('giConsole/count', entriesCount)
+
+  verifyLogsSize()
+
+  // To avoid infinite loop because we log all selector calls, we run sbp calls
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyLogsSize () {
+  if (entriesCount % ENTRIES_MARKER === 0) {
+    const markers = getMarkers()
+    // Save entry as a marker to be later deleted when logs are too big.
+    markers.push(lastEntry)
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+  }
+
+  if (entriesCount >= ENTRIES_LIMIT) {
+    const markers = getMarkers()
+    let toDelete = ENTRIES_MARKER // delete the latest x entries
+    let prevEntry = markers.shift()
+    do {
+      const entry = JSON.parse(localStorage.getItem(`giConsole/${prevEntry}`)) || {}
+      localStorage.removeItem(`giConsole/${prevEntry}`)
+      prevEntry = entry.prev
+      entriesCount--
+      toDelete--
+    } while (toDelete && prevEntry)
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+    localStorage.setItem('giConsole/count', entriesCount)
+  }
+}
+
+export function captureLogsStart () {
+  console.debug('captureLogsStart called')
+  isCapturing = true
+  lastEntry = localStorage.getItem('giConsole/lastEntry')
+  entriesCount = +localStorage.getItem('giConsole/count') || 0
```

Comment:
To make it so that we can adjust `ENTRIES_LIMIT`, can you make it so that when `captureLogsStart` is called, we do a real calculation for `entriesCount`?

I.e. check if `lastEntry` exists, and if so, start at `1` and loop backward through all the `.prev`s until either you get to a `localStorage.getItem('giConsole/${prev}')` that doesn't exist (incrementing `entriesCount` as you go). If `lastEntry` doesn't exist, set it to `0`.

Reply:
Why do you think that's needed?

When we set an entry to LS we also update `'giConsole/count'`. So, we must believe `count` value is correct when `captureLogsStart` is started the same way `lastEntry` is.

Reply:
It won't be correct if `ENTRIES_LIMIT` is changed. So we need to have code to recalculate the number each time `captureLogsStart` is called in case we want to be able to change `ENTRIES_LIMIT`

Reply:
if we change ENTRIES_LIMIT (imagine from 200 to 100 and there are already 110 logs stored) once a new log is captured it will verify there are 111 logs and it will perform the algorithm to clean the oldest logs...

Reply:
I tried this during my testing. If `ENTRIES_LIMIT` is changed then the `entriesCount` gets out of sync, which makes sense because it's simply being incremented and decremented. Give it a try yourself if you don't believe me :)

Reply:
Oh Greg... it's not about not believing in you. It's about understanding what you are saying and what's wrong with the code... I can't make a change if I don't understand that's the problem.

I'll take a closer look at it later and let you know when it's done

Reply:
Okay, done. Now when `captureLogsStart` is called, it `verifyLogsConfigs`. This will recalculate the markers array if `ENTRIES_MARKER_NTH` was changed and delete the oldest logs if `ENTRIES_LIMIT` was changed. Please take a look and give it a try on your computer.